keps/sig-autoscaling/20181106-in-place-update-of-pod-resources.md

```diff
@@ -0,0 +1,415 @@
+ +----------v---------+         |   |
+ |                    |         |   |
+ |        true        |         |   |
+ | AwaitingPreemption |         |   |
+ |                    |         |   |
+ +----------+---------+         |   |
+           2|                   |   |3
+            |           +-------v---+-------+
+            |           |                   |
+            +----------->       true        |
+                        | AwaitingPodResize |
+                        |                   |
+                        +-------------------+
+
+```
+
+1. Scheduler, on starting pre-emption.
+1. Scheduler, after pre-emption or no pre-emption needed.
+1. Kubelet, on completion of handling the resize request.
+
+The following diagram shows possible transitions of PodResizeSuccess
+PodCondition's Status and Reason fields respectively.
+
+```text
+
+                                  +---------+
+                                  |         |
+            +---------------------+  true   <------------+
+            |                     | Success |            |
+            |                     |         |            |
+            |                     +----+----+            |
+           1|                         2|                 |
+ +----------v---------+      +---------v--------+        |
+ |                    |      |                  |        |
+ |        false       |      |       false      |        |
+ | FailedNodeCapacity |      | Rejected<Reason> |        |
+ |                    |      |                  |        |
+ +----------+---------+      +------------------+        |
+           3|                                            |
+            |                     +----------+           |
+            |                     |          |           |
+            +--------------------->   false  +------------+
+                                  | Retrying | 4
+                                  |          |
+                                  +----------+
+
+```
+
+1. Scheduler or Kubelet, if not enough space on Node.
+1. Any Controller, on permanent issue.
+1. Initiating actor (controller), on retry.
+1. On retry and successful resizing.
+
+#### Container resource limit update ordering
+
+When in-place resize is desired for multiple Containers in a Pod, Kubelet updates
+resource limit for the Containers as detailed below:
+  1. If resource resizing results in net-increase of a resource type (CPU or Memory),
+     Kubelet first updates Pod-level cgroup limit for the resource type, and then
+     updates the Container resource limit.
+  1. If resource resizing results in net-decrease of a resource type, Kubelet first
+     updates the Container resource limit, and then updates Pod-level cgroup limit.
+  1. If resource update results in no net change of a resource type, only the Container
+     resource limits are updated.
+In all the above cases, Kubelet applies Container resource limit decreases before
+applying limit increases.
```

Comment:
if updating the container resource limit (through CRI I assume) fails, then is the cgroup reverted?

From a Windows standpoint there's no cgroup managed by the kubelet. 100% of this will need to be done in CRI.

Reply:
I feel the best course here is to [restart the container](https://github.com/kubernetes/enhancements/pull/686/files/11942431990c0e1a92b6dd811e3c2f170f90b4d8#diff-57e345a188df5653422eb0ed20bf9442R326) - imho this is last resort action. The rationale being if a limit update fails, there is possibility of a revert action failing as well if we were to try to rollback things. Is there a better alternative?

Reply:
@PatrickLang what do you mean no cgroup managed by kubelet ? All cgroup (or whatever windows equivalent)settings  should be modified by CRI, is that not the case ?

Reply:
its probably worth evaluating how the cri implementor should handle this for non-burstable resources like memory.  is the kubelet going to be responsible for inducing reclaim by setting limits closer to its usage but not actually at the resource allocated value?  if we just pass the resource allocated value to the cri, it will not be sufficient for us to use this api to induce reclaim, so i think i prefer a kubelet keeps a heuristic for how it handles scale down of burstable resources.

Reply:
"Restarting the container" can violate the `ResizePolicy` you set for the container. I'd prefer simply failing the resizing

Reply:
I'm also concerned about hard-to-define behavior when scaling down resources for CRI implementations.

Reply:
Restarting container policy seems to have a use-case where Java apps using Xmx flag wont be able to use the increased mem unless the app is restarted.

Could you please elaborate on the what might turn out unexpected on scale-down?